### ISSUE TITLE
Ninja triggers GLIBCXX assertion and crashes

### ISSUE THREAD

**Issue by ImmanuelHaffner** [ninja-build/ninja#2472]
Building one of my company's projects, I just had Ninja crash with

```plain
/usr/include/c++/14.1.1/bits/stl_vector.h|1130| std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = Node*; _Alloc = std::allocator<Node*>; reference = Node*&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.
|| Subprocess aborted
|| CMake Error: Generator: execution of make failed. Make command was: /usr/bin/ninja -j 14 <TARGETS REDACTED>
|| [Finished in 2 seconds with code 1]
```

I don't remember changing anything on the build configuration side. I am on Ninja version `1.12.1`. Compiling the same project in a new clone of the repository in a different paths works flawlessly. It used to work in the original path until now.

The error occurs in my `RelWithDebInfo` build located in `path/to/project/build/RelWithDebInfo`. A nearly identically configured `Debug` build in `path/to/project/build/Debug` builds just fine.

This issue might be related to https://github.com/ninja-build/ninja/issues/2309, which also reports the same assertion is triggered.

Running the build in the debugger points me to `deps_log.cc`:

```cpp
205│     if (is_deps) {
206│       assert(size % 4 == 0);
207│       int* deps_data = reinterpret_cast<int*>(buf);
208│       int out_id = deps_data[0];
209│       TimeStamp mtime;
210│       mtime = (TimeStamp)(((uint64_t)(unsigned int)deps_data[2] << 32) |
211│                           (uint64_t)(unsigned int)deps_data[1]);
212│       deps_data += 3;
213│       int deps_count = (size / 4) - 3;
214│
215│       Deps* deps = new Deps(mtime, deps_count);
216│       for (int i = 0; i < deps_count; ++i) {
217│         assert(deps_data[i] < (int)nodes_.size());
218├───────> assert(nodes_[deps_data[i]]);
219│         deps->nodes[i] = nodes_[deps_data[i]];
220│       }
221│
222│       total_dep_record_count++;
223│       if (!UpdateDeps(out_id, deps))
224│         ++unique_dep_record_count;
225│     } else {
```

Sadly, `i` is optimized out, but it's out of bounds.

**UPDATE:** After reconfiguring with `cmake -S . -B build/RelWithDebInfo -G Ninja --fresh ...` the problem persisted. After completely wiping `build/RelWithDebInfo` and regenerating the build configuration it is working again :man_shrugging:

**Comment by ImmanuelHaffner** [comment]
It seems this error is back:
```plain
ninja: Entering directory `build/RelWithDebInfo'
/usr/include/c++/14.1.1/bits/stl_vector.h:1130: std::vector<_Tp, _Alloc>::reference std::vector<_Tp, _Alloc>::operator[](size_type) [with _Tp = Node*; _Alloc = std::allocator<Node*>; reference
 = Node*&; size_type = long unsigned int]: Assertion '__n < this->size()' failed.
[1]    232138 IOT instruction (core dumped)  ninja -C build/RelWithDebInfo sql_tests planner_test
```
I have
```
$ ninja --version
1.12.1
```

**Comment by ImmanuelHaffner** [comment]
Hey @jhasse, do you know how much longer it will take you for the release of 1.13?
I am waiting for this patch. Could you release it sooner as a patch version of 1.12?

**Comment by jhasse** [comment]
1.13.0 will probably take quite a while (like a year or something). Doing 1.12.2 might be a good idea, I will think about it and see if I find the time.

**Comment by jhasse** [comment]
GitHub Actions no longer works on the release branch ... I don't have the time to fix that, so there won't be a 1.12.2 release unless someone steps up and finds a way to get it to work on RHEL 7.

I will try to hurry up with 1.13.0 instead.
